utils: redesign reusable_buffer

michoecho opened the pull request:
Common compression libraries work on contiguous buffers.
Contiguous buffers are a problem for the allocator. However, as long as they are short-lived,
we can avoid the expensive allocations by reusing buffers across tasks. 

This idea is already applied to the compression of CQL frames, but with some deficiencies.
`utils: redesign reusable_buffer` attempts to improve upon it in a few ways. See its commit message for an extended discussion.

Compression buffer reuse also happens in the zstd SSTable compressor, but the implementation is misguided. Every `zstd_processor` instance reuses a buffer, but each instance has its own buffer. This is very bad, because a healthy database might have thousands of concurrent instances (because there is one for each sstable reader). Together, the buffers might require gigabytes of memory, and the reuse actually *increases* memory pressure significantly, instead of reducing it.
`zstd: share buffers between compressor instances` aims to improve that by letting a single buffer be shared across all instances on a shard.

michoecho commented:
All proposed dynamics of `reusable_buffer` are up for discussion. For example, I don't have any maths behind the "definition" of a pathological request as 8x bigger than the max of the next period. To think about what is the probability of reallocation induced by that, we would need to know something about the distribution of request sizes.

The fixed decay periods are also arbitrary. I thought about making them self-adjusting (e.g. if a resize is repeated during some (big) time window, we double the decay period, and if it's noticed that for a long time no resize would have happened even with a halved decay period, then we halve it), but without a solid theoretical foundation it's most likely more complexity than the problem is worth. Even without it, the design might be already overengineered.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/295/

michoecho commented:
CI failed due to https://github.com/scylladb/scylladb/issues/13211. Restarting.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/299/

avikivity commented:
Please remove the link to the private repository, and add information as needed.

avikivity commented on test/boost/reusable_buffer_test.cc:
```diff
@@ -6,64 +6,72 @@
  * SPDX-License-Identifier: AGPL-3.0-or-later
  */
 
```
Also a reference in this patch's commit message. Please create a new issue to fix.

michoecho replied:
Created https://github.com/scylladb/scylladb/issues/13437

michoecho replied:
Done.

avikivity commented on utils/reusable_buffer.hh:
```diff
@@ -8,123 +8,156 @@
+            // In other words, we "define" a pathological request as one that is
+            // at least 8x bigger than the stable size, where the stable size
+            // is the max of all sizes seen within a decay period.
+            resize(_high_watermark);
```
I think we should shrink to 2*_high_watermark here. The logic is that is (say) the typical request size hovers round 256k - epsilon, we'd allocate 512k, not 256k, and catch small deviations.

Better to waste some memory than to cause unnecessary large allocations. They are very expensive!

I'd also raise the period to 10 minutes (no real justification).

michoecho replied:
> The logic is that is (say) the typical request size hovers round 256k - epsilon, we'd allocate 512k, not 256k, and catch small deviations.

Disagreed. The resistance to small deviations is already assured by the surrounding `if`: we shrink only if we haven't observed in the last period any allocations at least 1/16 as big as the current size. If a small oscillation (around a single power of 2) of the running max is happening, two large allocations may happen (iff the first allocation is smaller than the power of 2), but no more.

Rounding the size up to the *second* greater-or-equal power of 2 could seem to guarantee that *exactly* one allocation happens, instead of at most two. But this isn't true, because the initial allocations may be *2x smaller* than the running max, so they will cause the same reallocation anyway. So it only adds complexity without additional benefit.

> I'd also raise the period to 10 minutes (no real justification).

Ok, why not.

avikivity commented on utils/reusable_buffer.hh:
```diff
@@ -8,123 +8,156 @@
+ * The guard can be used to obtain a view at most once.
+ */
+template <typename Clock>
+class reusable_buffer {
```
To avoid over-templating here, let's make reusable_buffer a non-template and add a wrapper class that is a template on the clock, and only adds the timer functionality,

michoecho replied:
Done. I hate splitting the class just to make the job easier for the compiler, though.

avikivity commented:
See also: MREMAP_DONTUNMAP in [mremap](https://man7.org/linux/man-pages/man2/mremap.2.html). Unfortunately can't be assumed to be present everywhere. We'd need to use alien to call it, since it can't be called concurrently without huge stalls.

michoecho commented:
v2: Addressed Avi's comments.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/473/

michoecho commented:
I missed the failing CI, I thought it passed. It failed due to `ttl_test.TestDistributedTTL.test_ttl_is_respected_on_repair`. Restarting.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/560/

michoecho commented:
Pinging for review.

michoecho commented on utils/reusable_buffer.hh:
```diff
@@ -8,161 +8,311 @@
+        // is freed before allocating the new buffer.
+        // This lesses the pressure on the allocator and should guarantee
+        // a success if the new size is smaller than the old size.
+        _buf.clear();
```
> clear() doesn't deallocate the storage of the vector, it just runs the destructors of the elements.

Ouch, I was sure clear() deallocates. My life was a lie. Good catch, will fix.

michoecho replied:
Done.

michoecho commented on utils/reusable_buffer.hh:
```diff
@@ -8,161 +8,311 @@
+ *
+ * The reference/use counts exist only for checking purposes and should
+ * have no runtime effect (all guard code should be optimized out in a
+ * well-formed program).
```
I made the comment sound different than I wanted. I only meant to stress that the increments/decrements are only used for assert purposes and a correct program would behave exactly the same if you removed them. "Optimized out" is not the important part. I'll reword the comment.

michoecho replied:
Done.

michoecho commented:
v3: addressed Botond's comments:
- switched the underlying buffer from std::vector to std::unique_ptr<...[]>
- made sure that we actually free the old buffer before we allocate a new one (I wrongly only called vector::clear() before, which doesn't actually free memory)
- reworded the misleading comment about guard code being "optimized out"

michoecho commented:
If there are no other comments, please merge.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/906/

mykaul commented:
I reckon this is not backport material?

avikivity commented:
It's not.
